### ISSUE TITLE
Support Promises and Injected Services for Label Templates

### ISSUE THREAD

**Issue by Nicholas-Westby** [kgiszewski/Archetype#422]
There is a bit of documentation here about label templates: https://github.com/kgiszewski/ArchetypeManual/blob/master/02%20-%20Configuration.md

And there are a few sample label templates here: https://github.com/kgiszewski/Archetype/blob/abf6917b1cfb908dd4ce520beacee57638d3049c/app/helpers/sampleLabelHelpers.js

That works fine for simple functions that immediately return a value, but it gets a fair bit more complicated if you need to get a value from the server using the `$http` service or another custom AngularJS service. It would be nice if there were some means of using dependency injection to inject a service (like AngularJS controllers), and it would also be nice if promises could be returned from the function (the idea being that the promise would resolve to a string value).

To give you a sense of how harry this can get, here's some code I made to get the text version of a drop down multiple (i.e., it will display "Monday, Tuesday" instead of the "4565, 4566" prevalues):

```javascript
// Variables.
window.RhythmLabelTemplates = window.RhythmLabelTemplates || {};

// Don't pollute global scope.
(function () {

    // Create an AngularJS service to get prevalue text.
    var app = angular.module("umbraco");
    app.factory("rhythmPrevalues", function ($http) {

        // Variables.
        var services = {
            $http: $http
        };

        // Return service.
        return {
            getAndCacheValues: getGetAndCacheValues(services)
        };

    });

    /**
     * Returns the function that caches the prevalue text.
     * @param services The AngularJS services used by this function.
     * @returns {Function} The function that caches the prevalue text.
     */
    function getGetAndCacheValues(services) {

        // Variables.
        var $http = services.$http;

        // Return the function.
        return function (prevalues) {

            // Variables.
            var newPrevalues = [];
            var cachedPrevalues = window.RhythmLabelTemplates.DropDown.cachedPrevalues;

            // Figure out which prevalues have not yet been addressed.
            for (var i = 0; i < prevalues.length; i++) {
                var prevalue = prevalues[i];
                if (!cachedPrevalues[prevalue]) {
                    cachedPrevalues[prevalue] = "Loading...";
                    newPrevalues.push(prevalue);
                }
            }

            // Any new prevalues to get the text for?
            if (newPrevalues.length) {

                // Variables.
                var options = {
                    cache: false,
                    params: {
                        // Cache buster ensures requests aren't cached.
                        CacheBuster: Math.random(),
                        Prevalues: prevalues
                    }
                };

                // Ask the server for the prevalue text.
                var url = "/umbraco/backoffice/rhythm/prevalues/getvalues";
                $http.get(url, options)
                    .then(function (data) {

                        // Cache the text for each prevalue.
                        for (var i = 0; i < data.data.Items.length; i++) {
                            var pair = data.data.Items[i];
                            cachedPrevalues[pair.Key] = pair.Value;
                        }

                    });

            }

        };

    }

})();

/**
 * Returns a label template for a drop down so Archetype can display a nice preview.
 * @param value The drop down values.
 * @param scope Unused.
 * @param args Unused.
 * @returns {*} The transformed value.
 */
window.RhythmLabelTemplates.DropDown = function (value, scope, args) {

    // Variables.
    var me = window.RhythmLabelTemplates.DropDown;

    /**
     * Attempt to get the "rhythmPrevalues" AngularJS service.
     * @returns {*} The "rhythmPrevalues" AngularJS service, or null.
     */
    function getRhythmPrevaluesService() {
        try {
            return angular.element($(".umb-property").get(0)).injector().get("rhythmPrevalues");
        } catch(ex) {
            return null;
        }
    }

    // Get the text version of the prevalues.
    me.rhythmPrevalues = me.rhythmPrevalues || getRhythmPrevaluesService();
    if (value && value.length && me.rhythmPrevalues) {

        // Drop downs that store a single value are integers rather than arrays.
        if (!angular.isArray(value)) {
            value = [value.toString()];
        }

        // Get the text versions.
        var converted = [];
        me.rhythmPrevalues.getAndCacheValues(value);
        for (var i = 0; i < value.length; i++) {
            var prevalue = value[i];
            var label = me.cachedPrevalues[prevalue];
            converted.push(label || "Unknown");
        }

        // Combine the text values and separate with a comma.
        return converted.join(", ");

    }

    // Unable to get prevalue text.
    return "Unknown";

};

// Initialize cached variables.
window.RhythmLabelTemplates.DropDown.cachedPrevalues = {};
window.RhythmLabelTemplates.DropDown.rhythmPrevalues = null;
```

Here's what that looks like visually:

![office-hours](https://user-images.githubusercontent.com/5933222/29194316-d60ba160-7ddd-11e7-99d1-5a190769c92c.png)

There is some wonkiness in that code (side effect of me having no time and creating workarounds as I figured things out), but the short version is that this is more complicated than it needs to be.

Supporting injected services and promise return values would help simplify these label templates massively.

**Comment by kjac** [comment]
Hey @Nicholas-Westby,

Sounds like a nice addition. You're more than welcome to do a PR, if you're up for it.

**Comment by Nicholas-Westby** [comment]
Cool, I have a few ideas. I'll take a look next chance I get.

**Comment by kgiszewski** [comment]
Yea, looks impressive :)

**Comment by kjac** [comment]
@Nicholas-Westby awesome. I'll be keeping an eye out for it :)

**Comment by Nicholas-Westby** [comment]
Just a few notes for myself...

# Dependency Injection

This has a useful example of using dependency injection in custom scenarios: https://docs.angularjs.org/api/ng/function/angular.injector

Some code I tested from the console:

```javascript
angular.injector(["ng"]).invoke(function($timeout) {
  $timeout(function () {
    alert("hi");
  }, 2000);
});
```

Though, if you are already in a controller, you don't need to call `angular.injector` to get an instance of an `$injector`. You can just have it injected. Documentation on that service: https://docs.angularjs.org/api/auto/service/$injector

# Promises

If I want to check if a value is a promise, I can check it like so (`value && isFunction(value.then)`): https://stackoverflow.com/a/20723903

**Comment by Nicholas-Westby** [comment]
FYI, I'm doing a bit of work on this now. May not finish it this weekend, but I've got a good start.

**Comment by kgiszewski** [comment]
👍

**Comment by Nicholas-Westby** [comment]
Anybody know the significance of the pipe on this line: https://github.com/kgiszewski/Archetype/blob/master/app/services/archetypeLabelService.js#L275

```javascript
// split the template in case it consists of multiple property aliases and/or functions
var templates = match.replace("{{", '').replace("}}", '').split("|");
```

I see the comment, but I'm not entirely sure what it means. A use case and example might help clarify it.

**Comment by kgiszewski** [comment]
IIRC it allows more than one template to be passed.

**Comment by Nicholas-Westby** [comment]
I think there may be a bug in it (or maybe not), so I'm still working on trying to understand the intention of what it's supposed to be doing. Here's what I'm imagining so far. Suppose I have a label template like this:

```
Person Named {{firstName|lastName|middleInitial|CreateRandomFakeName()}}
```

Reading through the code, it seems like it's trying to find the first value that is returned and use that (rather than, say, concatenating the result of each value).

So, in my example, it'd first look for a first name. If that didn't have a value, it would look for a last name. If that didn't exist, it would look for the middle initial. If that didn't exist, it would call a function to generate a fake name to use. The end result would be a template that gets rendered something like this:

```
Person Named John
```

And it would NOT be something like this:

```
Person Named JohnSmithRMxyzptlk
```

Correct?

**Comment by kgiszewski** [comment]
I think that sounds right. It's been a long time since that code was written.

**Comment by Nicholas-Westby** [comment]
I got promises working: https://github.com/Nicholas-Westby/Archetype/tree/feature/422-label-template-enhancements

That means you can now create a label template function that does something like this:

```javascript
// Waits 2 seconds, then return the value wrapped in some spaces. Actually returns
// a promise immediately, but the promise will resolve to a string after 2 seconds.
function waitAndPad(value) {
    return $timeout(function () {
        return " " + value + " ";
    }, 2000);
}
```

Next up, dependency injection, which will allow for you to inject that `$timeout` service (or any service).

**Comment by Nicholas-Westby** [comment]
Pull request submitted: https://github.com/kgiszewski/Archetype/pull/425

Here are a few example label template functions that demonstrate the existing and new functionality:

```javascript
// Returns a string value.
function testString(value) {
    return "String: " + value;
}

// Returns a function to be invoked with dependency injection,
// when in turn returns a string value.
function testInjection(value) {
    return function($timeout) {
        var wasInjected = $timeout !== null;
        var prefix = wasInjected
            ? "Injected: "
            : "Not Injected: ";
        return prefix + value;
    }
}

// Returns a function to be invoked with dependency injection,
// which in turn returns a promise to be resolved to a string value.
function testPromise(value) {
    return function ($timeout) {
        return $timeout(function () {
            return "As Promised: " + value;
        }, 1000);
    }
}
```

The first one, `testString`, simply does what was already possible (i.e., returns a string).

The second one, `testInjection`, demonstrates that you can return a function and that that function will be called with the parameters injected.

The third one, `testPromise`, demonstrates the same thing as the second one, but it also returns a promise which gets resolved to a string.

While not necessarily recommended, this also technically supports other scenarios. For example, you can return a promise without a wrapper function for dependencies. Or you can have a function that returns a function that returns a promise that resolves to a function that returns a promise that resolves to a string, and so on.

In short, this gives the power of AngularJS's dependency injection system and the power of promises to label template functions.

The promises should come in handy with things like the `$http` service that are asynchronous and so don't return a value immediately. The dependency injection should come in  handy to get instances of services like `$http`.

**Comment by Nicholas-Westby** [comment]
BTW, I'll work on some documentation once this gets all approved and merged in.

**Comment by kgiszewski** [comment]
Fixed by #422
